src/conn.rs

```diff
@@ -0,0 +1,215 @@
+};
+use mentat_db::db;
+use mentat_db::{
+    transact,
+    PartitionMap,
+    TxReport,
+};
+use mentat_tx_parser;
+use query::{
+    q_once,
+    QueryResults,
+};
+
+
+/// Connection metadata required to query from, or apply transactions to, a Mentat store.
+///
+/// Owned data for the volatile parts (generation and partition map), and `Arc` for the infrequently
+/// changing parts (schema) that we want to share across threads.
+///
+/// See https://github.com/mozilla/mentat/wiki/Thoughts:-modeling-db-conn-in-Rust.
+pub struct Metadata {
+    pub generation: usize,
+    pub partition_map: PartitionMap,
+    pub schema: Arc<Schema>,
+}
+
+impl Metadata {
+    // Intentionally not public.
+    fn new(generation: usize, partition_map: PartitionMap, schema: Arc<Schema>) -> Metadata {
+        Metadata {
+            generation: generation,
+            partition_map: partition_map,
+            schema: schema,
+        }
+    }
+}
+
+/// A mutable, safe reference to the current Mentat store.
+struct Conn {
+    /// `Mutex` since all reads and writes need to be exclusive.  Internally, owned data for the
+    /// volatile parts (generation and partition map), and `Arc` for the infrequently changing parts
+    /// (schema) that we want to share across threads.  A consuming thread may use a shared
+    /// reference after the `Conn`'s `Metadata` has moved on.
+    ///
+    /// The motivating case is multiple query threads taking references to the current schema to
+    /// perform long-running queries while a single writer thread moves the metadata -- partition
+    /// map and schema -- forward.
+    metadata: Mutex<Metadata>,
+
+    // TODO: maintain set of change listeners or handles to transaction report queues. #298.
+
+    // TODO: maintain cache of query plans that could be shared across threads and invalidated when
+    // the schema changes. #315.
+}
+
+impl Conn {
+    // Intentionally not public.
+    fn new(partition_map: PartitionMap, schema: Schema) -> Conn {
+        Conn {
+            metadata: Mutex::new(Metadata::new(0, partition_map, Arc::new(schema)))
+        }
+    }
+
+    pub fn connect(sqlite: &mut rusqlite::Connection) -> Result<Conn> {
+        let db = db::ensure_current_version(sqlite)
+            .chain_err(|| "Unable to initialize Mentat store")?;
+        Ok(Conn::new(db.partition_map, db.schema))
+    }
+
+    /// Yield the current `Schema` instance.
+    pub fn current_schema(&self) -> Arc<Schema> {
+        // We always unwrap the mutex lock: if it's poisoned, this will propogate panics to all
+        // accessing threads.  This is perhaps not reasonable; we expect the mutex to be held for
+        // very short intervals, but a panic during a critical update section is possible, since the
+        // lock encapsulates committing a SQL transaction.
+        //
+        // That being said, in the future we will provide an interface to take the mutex, providing
+        // maximum flexibility for Mentat consumers.
```

Comment:
The only thing one can do in this case is recover, probably by effectively closing and reopening. File an issue?

Poisoning should be rare.